### PR TITLE
KAFKA-3055: Fix the JsonConverter mangling the Schema in connect

### PR DESCRIPTION
**Pull request by ksenji**

### CODE REVIEW

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java [hunk 1]
```diff
@@ -404,7 +404,7 @@ private ObjectNode asJsonSchema(Schema schema) {
                 jsonSchema = JsonNodeFactory.instance.objectNode().put(JsonSchema.SCHEMA_TYPE_FIELD_NAME, JsonSchema.STRUCT_TYPE_NAME);
                 ArrayNode fields = JsonNodeFactory.instance.arrayNode();
                 for (Field field : schema.fields()) {
-                    ObjectNode fieldJsonSchema = asJsonSchema(field.schema());
+                    ObjectNode fieldJsonSchema = asJsonSchema(field.schema()).deepCopy(); //KAFKA-3055
```
Comment (lindong28) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java]: We probably don't need to have this comment in the code.

**Comment by lindong28** [comment]
@ksenji Thanks for the patch. Can you add your testcase to pull request as well? That seems useful.

**Comment by ksenji** [comment]
@lindong28 Added the scenario to the existing test case: structToJson and removed the comment as well.

**Comment by lindong28** [comment]
@ksenji I am able to verify the problem described in the JIRA ticket. But the test case doesn't capture the problem, i.e. it still passes even without `deepCopy()`. Can you update the test case?

**Comment by ksenji** [comment]
@lindong28 Without the deepCopy(), I'm seeing the test case fail.

<pre><code>
java.lang.AssertionError: expected:<{"type":"struct","optional":false,"fields":[{"field":"field1","type":"boolean","optional":false},{"field":"field2","type":"string","optional":false},{"field":"field3","type":"string","optional":false},{"field":"field4","type":"boolean","optional":false}]}> but was:<{"type":"struct","fields":[{"type":"boolean","optional":false,"field":"field4"},{"type":"string","optional":false,"field":"field3"},{"type":"string","optional":false,"field":"field3"},{"type":"boolean","optional":false,"field":"field4"}],"optional":false}>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:118)
    at org.junit.Assert.assertEquals(Assert.java:144)
    at org.apache.kafka.connect.json.JsonConverterTest.structToJson(JsonConverterTest.java:438)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
    at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
</code></pre>

**Comment by lindong28** [comment]
@ksenji Yeah you are right. My ran the wrong test.

I am able to reproduce the problem and verify the fix. The patch LGTM. You may want to ping one committer for review and push the patch. Thanks.

**Comment by ksenji** [comment]
@ewencp Please review and merge this PR. The JIRA (https://issues.apache.org/jira/browse/KAFKA-3055) is assigned to you.

**Comment by ewencp** [comment]
LGTM. Thanks for the patch @ksenji and the review @lindong28!

**Pull request closed without merging**